server/sonar-db-dao/src/main/java/org/sonar/DBSessionsImpl.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar;
+
+import java.lang.reflect.Field;
+import java.util.Objects;
+import java.util.function.Supplier;
+import org.apache.ibatis.session.SqlSession;
+import org.apache.ibatis.session.defaults.DefaultSqlSession;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.db.DbSession;
+import org.sonar.db.MyBatis;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.lang.Thread.currentThread;
+
+public class DBSessionsImpl implements DBSessions {
+  private static final Logger LOG = Loggers.get(DBSessionsImpl.class);
+
+  private static final ThreadLocal<Boolean> CACHING_ENABLED = ThreadLocal.withInitial(() -> Boolean.FALSE);
+  private final ThreadLocal<DelegatingDbSessionSupplier> regularDbSession = ThreadLocal.withInitial(this::buildRegularDbSessionSupplier);
+  private final ThreadLocal<DelegatingDbSessionSupplier> batchDbSession = ThreadLocal.withInitial(this::buildBatchDbSessionSupplier);
+
+  private final MyBatis myBatis;
+
+  public DBSessionsImpl(MyBatis myBatis) {
+    this.myBatis = myBatis;
+  }
+
+  private DelegatingDbSessionSupplier buildRegularDbSessionSupplier() {
+    LOG.trace("{} called buildRegularDbSessionSupplier", currentThread());
```

Comment:
the point of this log is to have a way of, indeed, debugging in case anything fishy is going on with the threads, but as it's very verbose, I don't think debug is appropriate